[TorchTidy] Add pattern to detect if single-tensor implementation optimizer is used

davidchencsl opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81941
* #81921
* #81740
* __->__ #81733
* #81501

Summary: For Adam, SGD, AdamW, enabling multi-tensor implementation is faster. So for these optimizers, we detect if single-tensor implementation is used.

Test Plan:
Added test in test_profiler.py

Differential Revision: [D38082641](https://our.internmc.facebook.com/intern/diff/D38082641)

davidchencsl committed:
[TorchTidy] Add pattern to detect if single-tensor implementation optimizer is used

[ghstack-poisoned]

davidchencsl committed:
Update on "[TorchTidy] Add pattern to detect if single-tensor implementation optimizer is used"




[ghstack-poisoned]

robieta commented on test/test_profiler.py:
```diff
@@ -1582,6 +1583,37 @@ def test_profiler_extra_cuda_copy_pattern_benchmark(self):
+
+        num_matched = []
+        for _, fn in cases:
+            optimizer = torch.optim.Adam(model.parameters())
```
leftover?

robieta commented on test/test_profiler.py:
```diff
@@ -1582,6 +1583,37 @@ def test_profiler_extra_cuda_copy_pattern_benchmark(self):
 
+    def test_profiler_optimizer_single_tensor_pattern(self):
+        x = torch.ones((100, 100))
+        model = nn.Sequential(
```
You'll want to make a new model each time since the optimizer will be coupled to the model. (Shouldn't matter, but might as well.)

davidchencsl replied:
👍

robieta reviewed (approved):
Fantastic

davidchencsl committed:
Update on "[TorchTidy] Add pattern to detect if single-tensor implementation optimizer is used"


Summary: For Adam, SGD, AdamW, enabling multi-tensor implementation is faster. So for these optimizers, we detect if single-tensor implementation is used.

Test Plan:
Added test in test_profiler.py

[ghstack-poisoned]

robieta reviewed (approved):
LGTM

davidchencsl committed:
Update on "[TorchTidy] Add pattern to detect if single-tensor implementation optimizer is used"


Summary: For Adam, SGD, AdamW, enabling multi-tensor implementation is faster. So for these optimizers, we detect if single-tensor implementation is used.

Test Plan:
Added test in test_profiler.py

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082641).

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082641).

davidchencsl commented:
@pytorchbot merge

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082641).
